### ISSUE TITLE
Add Support for multiple types of connectors to Magento 2 instance

### ISSUE THREAD

**Issue by PWAStudioBot** [magento-research/pwa-studio#28]
_From @Jakhotiya on May 14, 2018 13:42_

This repository doesn't work well without vagrant. We need to add support for trying out this code without needing vagrant. On my local I had to fix many url problems to get venia work. 

I would like to help but can somebody point me in right direction

_Copied from original issue: magento-research/venia-pwa-concept#70_

**Comment by PWAStudioBot** [comment]
_From @DrewML on May 14, 2018 15:7_

Hi @Jakhotiya - appreciate the feedback. Can you outline what problems you had attempting to set this up without Vagrant?

The (admittedly sparse) setup docs go through the Vagrant setup process since that seems to be the most common among m2 users, but most of us on the team are using this with Valet+ on our hosts.

**Comment by PWAStudioBot** [comment]
_From @Jakhotiya on May 14, 2018 17:47_

### Environment setup
magento 2.3-develop branch
nodejs v9
nginx php7.1 fpm mysql stack
store url on my local 23-develop.to (Couldn't use url of the dev server provided with this repo)

### Setting up the module and theme
First I downloaded the code from github and moved the module to app/code/Magento and venia theme
to app/design/frontend/Magento/. Then I ran npm install in theme directory to setup the dependencies. I ran npm start to start up the dev server. This generated client.js and other javascript files as expected. After stopping dev server I ran setup:upgrade to install the theme and the pwa module. I activated venia theme. After opening my store front these are issues I had to fix
1) **404 for client.js** (because the path was /pub/static/frontend/Magento/venia/en_US/client.js instead of /pub/static/frontend/Magento/venia/en_US/js/client.js)
2) client.js:1 Uncaught ReferenceError: webpackJsonp is not defined
    at client.js:1. To fix this I had to include vendor.js before client.js
3) Everything loads but screen remains black with "loading" text. I noticed response from http://23-develop.to/graphql an request was 
```
{"data":{"urlResolver":null}}
```
It would be great if we could move the module-pwa in a separate repository as pointed out in [issue](https://github.com/magento-research/venia-pwa-concept/issues/62)

I'm still working on the last issue. But overall I would love to streamline these things and document them.

**Comment by PWAStudioBot** [comment]
_From @mujahid-emb on May 15, 2018 5:24_

Hi @Jakhotiya - Thanks for bringing up the issues. We also faced the issues 1 & 2 and resolved but stuck on the issue 3(screen shows "Loading" text). We are trying without Vagrant. Please let us know if you fix the issue 3.

**Comment by PWAStudioBot** [comment]
_From @leon0v on May 15, 2018 7:19_

@Jakhotiya @mujahid-emb 

> 3. Everything loads but screen remains black with "loading" text. I noticed response from http://23-develop.to/graphql an request was
{"data":{"urlResolver":null}}

There is still a [bug](https://github.com/magento-research/venia-pwa-concept/issues/66) that when route is "/" urlResolver returns null. Maybe this is the case with you too.

In this case try the route "/home" instead

**Comment by PWAStudioBot** [comment]
_From @Jakhotiya on May 15, 2018 7:37_

Thanks @leon0v That does help in changing the response received from server. Now instead of null it is
```
{"data":{"urlResolver":{"type":"CMS_PAGE"}}}
```
But interestingly That did not change anything in the UI. I still see "Loading" text. Any thoughts on that?

**Comment by PWAStudioBot** [comment]
_From @leon0v on May 15, 2018 7:48_

@Jakhotiya What does the console in your browser output? I also posted a temporarily fix below the [issue](https://github.com/magento-research/venia-pwa-concept/issues/66#issuecomment-389078878).

**Comment by PWAStudioBot** [comment]
_From @Jakhotiya on May 15, 2018 8:41_

Console says
Routing resolve failed
 TypeError: Failed to construct 'URL': Invalid base URL
I will try you fix.

**Comment by PWAStudioBot** [comment]
_From @leon0v on May 15, 2018 8:46_

What parameters do you use when initializing a new Peregrine() in your index.js?

Mine are:

`const app = new Peregrine({
    apiBase: new URL('/graphql', location.origin).toString(),
    __tmp_webpack_public_path__: __webpack_public_path__ + "/en_US"
});`

This error implies that you are using an invalid apiBase URL.

**Comment by PWAStudioBot** [comment]
_From @Jakhotiya on May 15, 2018 9:31_

My code used to match code base
```
const app = new Peregrine({
    apiBase: new URL('/graphql', location.origin).toString(),
    __tmp_webpack_public_path__: __webpack_public_path__
});
```

After reading your comment I tried 

```
const app = new Peregrine({
    apiBase: new URL('/graphql', location.origin).toString(),
    __tmp_webpack_public_path__: __webpack_public_path__ + "/en_US"
});
```
That changes error to Routing resolve failed
 Error: 404. Let me take a look at this and I should be able to come up with an answer. I would like to point out that I'm not using webpack development server. I'm using regular magento 2 url on port 80

**Comment by PWAStudioBot** [comment]
_From @DrewML on May 15, 2018 16:24_

@Jakhotiya because we're still very much in early active development on the theme, there hasn't been much of any focus on the deployment or production version of things yet. Right now the happy path is very much running the development server.

------

> 404 for client.js (because the path was /pub/static/frontend/Magento/venia/en_US/client.js instead of /pub/static/frontend/Magento/venia/en_US/js/client.js)

I'm sure we can address this in #63 

--------

> client.js:1 Uncaught ReferenceError: webpackJsonp is not defined
at client.js:1. To fix this I had to include vendor.js before client.js

This will be handled automatically when we focus more on the production build story. You don't see this in development because we don't split out a vendor chunk (since there isn't a need for cache granularity in dev).

-------

> That changes error to Routing resolve failed
Error: 404. Let me take a look at this and I should be able to come up with an answer. I would like to point out that I'm not using webpack development server. I'm using regular magento 2 url on port 80

If you can't figure this one out, please open up a new issue with a screenshot of the console in DevTools and we'll help debug.

-------

None of the issues you've outlined seem related to using a Vagrant setup, though. Seems like the root of most of these issues is the attempt to run the "production" build of the frontend. Is there a specific reason you don't wish to use the dev server at the moment?

**Comment by PWAStudioBot** [comment]
_From @Jakhotiya on May 15, 2018 19:14_

Thank you very much @DrewML for the explanation and a hint. I got dev server to work. I had forgotten to change host url in .env and kept getting proxying error(cryptic). After running dev server I do not have to worry about vendor.js. Indeed it has nothing to do with vagrant.

I noticed even after having my dev server running  javascript file requests kept on going to 404. I looked in pub/static/frontend/Magento/venia/en_US/ but the file did not exist. I was confused, because I thought webpack-dev server would generate such an file named client.js . **Only after much digging did I realize that dev server will keep all assets in memory hence there would be no physical file called client.js** 

One reason for such confusion was `/pub/static/frontend/Magento/venia/en_US/client.js` file path in `root.phtml`. Regular magento dev would think that this is a mistake. Especially if you run a build, you notice that client.js is correctly generated at path `/pub/static/frontend/Magento/venia/en_US/js/client.js`. My question is could we configure dev server to serve `/pub/static/frontend/Magento/venia/en_US/js/client.js` instead of `/pub/static/frontend/Magento/venia/en_US/client.js` and If I create pull request for such an change would it be acceptable?

I would also like to document all of these gotcha's of setup. Where would I put that?

**Comment by zetlen** [comment]
Will be resolved by #248 .

**Comment by ericerway** [comment]
Closed with #248. Cleaning up older stories in backlog post-2.3.0
